### DISCUSSION TITLE
Setting/getting size

### DISCUSSION THREAD

**Discussion by davidbrochart** [Textualize/textual#2200] [Q&A] [answered]
I'm trying to set the size of the terminal, but the application takes the size of my terminal anyway. Also, printing the size of the widget shows `Size(width=0, height=0)`:
```py
from textual.app import App, ComposeResult
from textual.widgets import Static


class MyStatic(Static):
    def on_mount(self):
        print(self.size)


class MyApp(App):

    def compose(self) -> ComposeResult:
        yield MyStatic("\n".join([f"{i} " * 100 for i in range(100)]))


if __name__ == "__main__":
    app = MyApp()
    app.run(size=(10, 10))
```

**Comment by willmcgugan** [comment, marked as answer]
The `size` argument was mainly for situations where there isn't a physical terminal, like testing. We could make that clearer in the docs.

The widget size will be (0, 0) on mount because you get the mount prior to the first layout.

What behaviour were you expecting by setting the size on run?

**Reply by davidbrochart** [reply]
I'm trying to reproduce an issue with a container that has a `Tabs` and a `Static`, and the `Static` seems to have a wrong height (it includes the height of the `Tabs`).

**Reply by willmcgugan** [reply]
Could you reproduce it by putting everything inside another container with size 10x10 ?

**Reply by davidbrochart** [reply]
Yes. In the following example, the container has a height of 10, but the widget's height is 20, although I can see that it displays correctly (only the first 7 rows are shown, since the height of the tabs is 3). But maybe the `size` property of a widget is its "full size", regardless of the container it is in?
```py
import asyncio

from textual.app import App, ComposeResult
from textual.containers import Container
from textual.widgets import Static, Tabs


class MyContainer(Container):

    def compose(self) -> ComposeResult:
        self.widget = Static("\n".join([f"{i} " * 20 for i in range(20)]))
        yield Tabs("title")
        yield self.widget
        asyncio.create_task(self.print_widget_size())

    async def print_widget_size(self):
        await asyncio.sleep(0.1)
        print(self.widget.size)


class MyApp(App):

    DEFAULT_CSS = """
    MyContainer {
        height: 10;
        width: 100%;
    }
    """

    def compose(self) -> ComposeResult:
        yield MyContainer()


if __name__ == "__main__":
    app = MyApp()
    app.run()
```

**Reply by willmcgugan** [reply]
> But maybe the size property of a widget is its "full size", regardless of the container it is in?

That's right. It's not the size on screen, but the size within its container.

**Reply by davidbrochart** [reply]
So how would I get the size available to the widget? Here I'm expecting a height of 7, since it's in a container of height 10, and 3 rows are taken by the tabs.

**Reply by willmcgugan** [reply]
I have a feeling this may be an XY problem. Can we step back and help me understand what you are trying to accomplish?

**Reply by davidbrochart** [reply]
I'm trying to have a container with two widgets aligned vertically, a `Tabs` and a `Static`. In my particular case, the `Static` will show a terminal emulator, so I need to get the available height of the `Static` and report it back to the backend. But it seems that the reported height includes the 3 rows of the `Tabs`.

**Reply by willmcgugan** [reply]
You can set the height of your widget to `1fr`, which will tell it to fill the remaining space. If the container has height 10, and the tables take up 3 rows, the widget will get the height of 7.

You could then implement `on_size` on your Widget to get the size when it changes.

**Reply by davidbrochart** [reply]
Great, thanks!
